architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+| ------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Resource Namespace | A group of related or similar resources that share a resource provider. eg: In `Applications.Core/containers` the resource namespace is `Applications.Core`.                                                                                                                                 |
+| Resource Provider  | A resource namespace and implementation of the resource namespace. Resource providers are a new resource type introduced by this proposal.                                                                                                                                                   |
+| Resource Type      | An API entity with CRUDL operations. Resource Types are part of a resource provider. Resource types are a new resource type introduced by this proposal.                                                                                                                                     |
+| OpenAPI/Schema     | A format for documenting HTTP-based APIs. We use OpenAPI v2 currently in Radius for client-sdk generation and validation of inbound requests. OpenAPI v3 is more widely adopted. In this document we're primarily concerned with the parts of OpenAPI pertaining to request/response bodies. |
+
+## Objectives
+
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6688
+
+### Goals
+
+- Support granular registration of resource types (add/change/remove) by users.
+- Support efficient querying of resource type data for any API consumer.
+- Register resource types for existing Radius functionality like Deployment Engine and `Applications.*` resource types.
+
+### Non goals
+
+- (out of scope): Complete the user-defined type implementation. This design document is scoped to registration of resource types.
+- (out of scope): Tooling for resource type registration (CLI, Bicep). This work is tracked by another design document.
+- (non-goal): Provider non-user-facing mechanisms for resource type registration.
+
+### User scenarios (optional)
+
+#### Defining new types
+
+Suzy is a platform engineer at Contoso working on the internal developer platform. She defines a resource type for an internal service `ContosoBus` named `Contoso.Platform/contosoBuses`. 
+
+She can use tools like Bicep, GitOps, or the `rad` CLI to register the new resource types in any tenant of Radius.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+#### Removing default resource types
```

Comment:
Is a more probabilistic use case to unregister custom resource types than default types?

Reply:
I think they are both pretty important. I went with the "unregister built-in stuff" as a scenario because it's less obvious. 

We've already validated with users that this is important to them.